ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/permissioning/NodeWhitelistController.java

```diff
@@ -45,45 +59,65 @@ private boolean removeNode(final Peer node) {
     return nodesWhitelist.remove(node);
   }
 
-  public NodesWhitelistResult addNodes(final List<DefaultPeer> peers) {
+  public NodesWhitelistResult addNodes(final List<Peer> peers) {
     final NodesWhitelistResult inputValidationResult = validInput(peers);
     if (inputValidationResult.result() != WhitelistOperationResult.SUCCESS) {
       return inputValidationResult;
     }
 
-    for (DefaultPeer peer : peers) {
+    for (Peer peer : peers) {
       if (nodesWhitelist.contains(peer)) {
         return new NodesWhitelistResult(
             WhitelistOperationResult.ERROR_EXISTING_ENTRY,
             String.format("Specified peer: %s already exists in whitelist.", peer.getId()));
       }
     }
+
+    final List<Peer> oldWhitelist = new ArrayList<>(this.nodesWhitelist);
+
     peers.forEach(this::addNode);
+    try {
+      updateConfigurationFile(peerToEnodeURI(nodesWhitelist));
+    } catch (IOException e) {
+      revertState(oldWhitelist);
+      return new NodesWhitelistResult(
+          WhitelistOperationResult.ERROR_WHITELIST_PERSIST_FAIL, WHITELIST_FAIL_MESSAGE);
+    }
     return new NodesWhitelistResult(WhitelistOperationResult.SUCCESS);
   }
 
-  private boolean peerListHasDuplicates(final List<DefaultPeer> peers) {
+  private boolean peerListHasDuplicates(final List<Peer> peers) {
     return !peers.stream().allMatch(new HashSet<>()::add);
   }
 
-  public NodesWhitelistResult removeNodes(final List<DefaultPeer> peers) {
+  public NodesWhitelistResult removeNodes(final List<Peer> peers) {
     final NodesWhitelistResult inputValidationResult = validInput(peers);
     if (inputValidationResult.result() != WhitelistOperationResult.SUCCESS) {
       return inputValidationResult;
     }
 
-    for (DefaultPeer peer : peers) {
+    for (Peer peer : peers) {
       if (!(nodesWhitelist.contains(peer))) {
         return new NodesWhitelistResult(
             WhitelistOperationResult.ERROR_ABSENT_ENTRY,
             String.format("Specified peer: %s does not exist in whitelist.", peer.getId()));
       }
     }
+
+    final List<Peer> oldWhitelist = new ArrayList<>(this.nodesWhitelist);
+
     peers.forEach(this::removeNode);
+    try {
+      updateConfigurationFile(peerToEnodeURI(nodesWhitelist));
+    } catch (IOException e) {
+      revertState(oldWhitelist);
```

Comment:
I think that we aren't ensuring that writing to the file is an atomic operation. In this case, I believe the best thing would be comparing the in-memory old list with the latest list read from the file to verify that they are the same. If they are not, it means that the file has been compromised.

Reply:
Updated.